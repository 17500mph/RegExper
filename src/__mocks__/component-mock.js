const React = require('react');

const buildMock = component => {
  const componentName = component.displayName || component.name || 'Component';
  const Mock = ({ children, ...props }) => (
    <span
      data-component={ componentName }
      data-props={ JSON.stringify(props, null, '  ') }>{ children }</span>
  );
  Mock.propTypes = component.propTypes;
  return Mock;
};

module.exports = path => {
  const actual = jest.requireActual(path).default;
  return buildMock(actual);
};

module.exports.buildMock = buildMock;
