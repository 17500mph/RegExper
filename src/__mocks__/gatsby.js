const React = require('react');
const gatsby = jest.requireActual('gatsby');

module.exports = {
  ...gatsby,
  graphql: jest.fn().mockImplementation(([query]) => query),
  Link: jest.fn().mockImplementation(({ to, ...rest }) =>
    React.createElement('a', {
      ...rest,
      href: to
    })
  ),
  StaticQuery: jest.fn()
};
