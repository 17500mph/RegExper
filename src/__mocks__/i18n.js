const i18n = jest.requireActual('i18n');

// Load empty resource bundle to reduce logging output
i18n.default.addResourceBundle('dev', 'translation', {});
i18n.default.addResourceBundle('en', 'translation', {});
i18n.default.addResourceBundle('other', 'translation', {});

module.exports = {
  ...i18n,
  locales: [
    { code: 'en', name: 'English' },
    { code: 'other', name: 'Other' }
  ],
  mockT: str => `TRANSLATE(${ str })`
};
