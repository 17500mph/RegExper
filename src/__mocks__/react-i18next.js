const reactI18next = jest.requireActual('react-i18next');

module.exports = {
  ...reactI18next,
  Trans: require('__mocks__/component-mock').buildMock(reactI18next.Trans)
};
