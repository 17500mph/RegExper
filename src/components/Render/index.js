import React from 'react';
import PropTypes from 'prop-types';

import nodeTypes from 'rendering/types';

import style from './style.module.css';

const render = (data, extraProps) => {
  if (typeof data === 'string') {
    return data;
  }

  const { type, props } = data;
  const children = (data.children || []).map(
    (data, key) => render(data, { key }));

  return React.createElement(
    nodeTypes[type] ? nodeTypes[type].default : type,
    { ...extraProps, ...props },
    children.length === 1 ? children[0] : children);
};

class Render extends React.PureComponent {
  static propTypes = {
    data: PropTypes.object.isRequired,
    onRender: PropTypes.func.isRequired
  }

  svgContainer = React.createRef()

  componentDidMount() {
    this.provideSVGData();
  }

  componentDidUpdate() {
    this.provideSVGData();
  }

  provideSVGData() {
    if (!this.svgContainer.current) {
      return;
    }

    const svg = this.svgContainer.current.querySelector('svg');
    this.props.onRender({
      svg: svg.outerHTML,
      width: Number(svg.getAttribute('width')),
      height: Number(svg.getAttribute('height'))
    });
  }

  render() {
    const { data } = this.props;

    return <div className={ style.render } ref={ this.svgContainer }>
      { render(data, { onReflow: this.provideSVGData }) }
    </div>;
  }
}

export default Render;
