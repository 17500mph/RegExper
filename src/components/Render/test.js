jest.mock('rendering/Box', () =>
  require('__mocks__/component-mock')('rendering/Box'));
jest.mock('rendering/HorizontalLayout', () =>
  require('__mocks__/component-mock')('rendering/HorizontalLayout'));
jest.mock('rendering/Loop', () =>
  require('__mocks__/component-mock')('rendering/Loop'));
jest.mock('rendering/Pin', () =>
  require('__mocks__/component-mock')('rendering/Pin'));
jest.mock('rendering/Text', () =>
  require('__mocks__/component-mock')('rendering/Text'));
jest.mock('rendering/VerticalLayout', () =>
  require('__mocks__/component-mock')('rendering/VerticalLayout'));

import React from 'react';
import { render } from 'react-testing-library';

import Render from 'components/Render';

const testType = (name, item) => {
  test(name, () => {
    const data = { type: 'SVG', children: [item] };
    const { asFragment } = render(
      <Render data={ data } onRender={ jest.fn() }/>
    );
    expect(asFragment()).toMatchSnapshot();
  });
};

describe('Render', () => {
  test('debugging', () => {
    const data = {
      type: 'SVG',
      children: [
        {
          type: 'Text',
          debug: true,
          box: {
            width: 100,
            height: 50,
            axisY: 10,
            axisX1: 5,
            axisX2: 95
          },
          children: [
            'Example'
          ]
        }
      ]
    };
    const { asFragment } = render(
      <Render data={ data } onRender={ jest.fn() }/>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('types', () => {
    testType('Pin', {
      type: 'Pin'
    });

    testType('Text', {
      type: 'Text',
      children: [
        'Example'
      ]
    });

    testType('Box', {
      type: 'Box',
      children: [
        {
          type: 'Text',
          children: [
            'Example'
          ]
        }
      ]
    });

    testType('Loop', {
      type: 'Loop',
      children: [
        {
          type: 'Text',
          children: [
            'Example'
          ]
        }
      ]
    });

    testType('HorizontalLayout', {
      type: 'HorizontalLayout',
      children: [
        {
          type: 'Text',
          children: [
            'Example'
          ]
        },
        {
          type: 'Text',
          children: [
            'Another Example'
          ]
        }
      ]
    });

    testType('VerticalLayout', {
      type: 'VerticalLayout',
      children: [
        {
          type: 'Text',
          children: [
            'Example'
          ]
        },
        {
          type: 'Text',
          children: [
            'Another Example'
          ]
        }
      ]
    });
  });
});
