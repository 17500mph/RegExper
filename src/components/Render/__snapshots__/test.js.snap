// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render debugging 1`] = `
<DocumentFragment>
  <div
    class="render"
  >
    <svg
      style="background-color: rgb(255, 255, 255);"
      viewBox="0 0  "
      xmlns="http://www.w3.org/2000/svg"
      xmlns:cc="http://creativecommons.org/ns#"
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
      <metadata>
        <rdf:rdf>
          
  
          <cc:license
            rdf:about="http://creativecommons.org/licenses/by/3.0/"
          >
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Reproduction"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Distribution"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Notice"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Attribution"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#DerivativeWorks"
            />
            
  
          </cc:license>
          

        </rdf:rdf>
      </metadata>
      <g
        transform="translate(10 10)"
      >
        <span
          data-component="Text"
          data-props="{}"
        >
          Example
        </span>
        <rect
          height="50"
          style="fill: transparent; stroke: red; stroke-width: 1px; stroke-dasharray: 2,2; opacity: 0.5;"
          width="100"
        />
        <circle
          cx="5"
          cy="10"
          r="3"
          style="fill: red; opacity: 0.5;"
        />
        <circle
          cx="95"
          cy="10"
          r="3"
          style="fill: red; opacity: 0.5;"
        />
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Render types Box 1`] = `
<DocumentFragment>
  <div
    class="render"
  >
    <svg
      style="background-color: rgb(255, 255, 255);"
      viewBox="0 0  "
      xmlns="http://www.w3.org/2000/svg"
      xmlns:cc="http://creativecommons.org/ns#"
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
      <metadata>
        <rdf:rdf>
          
  
          <cc:license
            rdf:about="http://creativecommons.org/licenses/by/3.0/"
          >
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Reproduction"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Distribution"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Notice"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Attribution"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#DerivativeWorks"
            />
            
  
          </cc:license>
          

        </rdf:rdf>
      </metadata>
      <g
        transform="translate(10 10)"
      >
        <span
          data-component="Box"
          data-props="{}"
        >
          <span
            data-component="Text"
            data-props="{}"
          >
            Example
          </span>
        </span>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Render types HorizontalLayout 1`] = `
<DocumentFragment>
  <div
    class="render"
  >
    <svg
      style="background-color: rgb(255, 255, 255);"
      viewBox="0 0  "
      xmlns="http://www.w3.org/2000/svg"
      xmlns:cc="http://creativecommons.org/ns#"
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
      <metadata>
        <rdf:rdf>
          
  
          <cc:license
            rdf:about="http://creativecommons.org/licenses/by/3.0/"
          >
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Reproduction"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Distribution"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Notice"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Attribution"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#DerivativeWorks"
            />
            
  
          </cc:license>
          

        </rdf:rdf>
      </metadata>
      <g
        transform="translate(10 10)"
      >
        <span
          data-component="HorizontalLayout"
          data-props="{}"
        >
          <span
            data-component="Text"
            data-props="{}"
          >
            Example
          </span>
          <span
            data-component="Text"
            data-props="{}"
          >
            Another Example
          </span>
        </span>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Render types Loop 1`] = `
<DocumentFragment>
  <div
    class="render"
  >
    <svg
      style="background-color: rgb(255, 255, 255);"
      viewBox="0 0  "
      xmlns="http://www.w3.org/2000/svg"
      xmlns:cc="http://creativecommons.org/ns#"
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
      <metadata>
        <rdf:rdf>
          
  
          <cc:license
            rdf:about="http://creativecommons.org/licenses/by/3.0/"
          >
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Reproduction"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Distribution"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Notice"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Attribution"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#DerivativeWorks"
            />
            
  
          </cc:license>
          

        </rdf:rdf>
      </metadata>
      <g
        transform="translate(10 10)"
      >
        <span
          data-component="Loop"
          data-props="{}"
        >
          <span
            data-component="Text"
            data-props="{}"
          >
            Example
          </span>
        </span>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Render types Pin 1`] = `
<DocumentFragment>
  <div
    class="render"
  >
    <svg
      style="background-color: rgb(255, 255, 255);"
      viewBox="0 0  "
      xmlns="http://www.w3.org/2000/svg"
      xmlns:cc="http://creativecommons.org/ns#"
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
      <metadata>
        <rdf:rdf>
          
  
          <cc:license
            rdf:about="http://creativecommons.org/licenses/by/3.0/"
          >
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Reproduction"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Distribution"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Notice"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Attribution"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#DerivativeWorks"
            />
            
  
          </cc:license>
          

        </rdf:rdf>
      </metadata>
      <g
        transform="translate(10 10)"
      >
        <span
          data-component="Pin"
          data-props="{}"
        />
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Render types Text 1`] = `
<DocumentFragment>
  <div
    class="render"
  >
    <svg
      style="background-color: rgb(255, 255, 255);"
      viewBox="0 0  "
      xmlns="http://www.w3.org/2000/svg"
      xmlns:cc="http://creativecommons.org/ns#"
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
      <metadata>
        <rdf:rdf>
          
  
          <cc:license
            rdf:about="http://creativecommons.org/licenses/by/3.0/"
          >
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Reproduction"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Distribution"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Notice"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Attribution"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#DerivativeWorks"
            />
            
  
          </cc:license>
          

        </rdf:rdf>
      </metadata>
      <g
        transform="translate(10 10)"
      >
        <span
          data-component="Text"
          data-props="{}"
        >
          Example
        </span>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Render types VerticalLayout 1`] = `
<DocumentFragment>
  <div
    class="render"
  >
    <svg
      style="background-color: rgb(255, 255, 255);"
      viewBox="0 0  "
      xmlns="http://www.w3.org/2000/svg"
      xmlns:cc="http://creativecommons.org/ns#"
      xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    >
      <metadata>
        <rdf:rdf>
          
  
          <cc:license
            rdf:about="http://creativecommons.org/licenses/by/3.0/"
          >
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Reproduction"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#Distribution"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Notice"
            />
            
    
            <cc:requires
              rdf:resource="http://creativecommons.org/ns#Attribution"
            />
            
    
            <cc:permits
              rdf:resource="http://creativecommons.org/ns#DerivativeWorks"
            />
            
  
          </cc:license>
          

        </rdf:rdf>
      </metadata>
      <g
        transform="translate(10 10)"
      >
        <span
          data-component="VerticalLayout"
          data-props="{}"
        >
          <span
            data-component="Text"
            data-props="{}"
          >
            Example
          </span>
          <span
            data-component="Text"
            data-props="{}"
          >
            Another Example
          </span>
        </span>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;
